server/proof_fetcher_test.go

```diff
@@ -0,0 +1,360 @@
+		nodes:   []storage.Node{sn1, sn2, sn3, sn4, sn5},
+		fetches: []merkle.NodeFetch{{Rehash: false}, {Rehash: true}, {Rehash: true}, {Rehash: true}, {Rehash: false}},
+		output: trillian.Proof{
+			LeafIndex: int64(23),
+			ProofNode: []*trillian.Node{n1, n2n3n4, n5},
+		},
+	},
+	{
+		desc:    "multiple rehash",
+		index:   int64(45),
+		nodes:   []storage.Node{sn1, sn2, sn3, sn4, sn5},
+		fetches: []merkle.NodeFetch{{Rehash: true}, {Rehash: true}, {Rehash: false}, {Rehash: true}, {Rehash: true}},
+		output: trillian.Proof{
+			LeafIndex: int64(45),
+			ProofNode: []*trillian.Node{n1n2, n3, n4n5},
+		},
+	},
+}
+
+// Test data used to test node fech deduplication in isolation
+type dedupTest struct {
+	desc         string
+	input        []merkle.NodeFetch // input fetches to deduper
+	storageIDs   []storage.NodeID   // ids sent to storage layer
+	storageNodes []storage.Node     // nodes returned by storage layer
+	storageError error              // error returned by storage layer
+	result       []storage.Node     // expected result with dupes reinstated
+}
+
+// Contents of these don't really matter as long as they have distinct IDs
+var f11 = merkle.NodeFetch{NodeID: mustCreateNodeID(1, 1)}
+var f12 = merkle.NodeFetch{NodeID: mustCreateNodeID(1, 2)}
+var f31 = merkle.NodeFetch{NodeID: mustCreateNodeID(3, 1)}
+
+var n11 = storage.Node{NodeID: f11.NodeID, NodeRevision: 37}
+var n12 = storage.Node{NodeID: f12.NodeID, NodeRevision: 37}
+var n31 = storage.Node{NodeID: f31.NodeID, NodeRevision: 37}
+
+var dedupTests = []dedupTest{
+	{
+		desc:         "no dupes",
+		input:        []merkle.NodeFetch{f11, f12, f31},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{n11, n12, n31},
+		result:       []storage.Node{n11, n12, n31},
+	},
+	{
+		desc:         "one dupe",
+		input:        []merkle.NodeFetch{f11, f12, f31, f11},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{n11, n12, n31},
+		result:       []storage.Node{n11, n12, n31, n11},
+	},
+	{
+		desc:         "multi dupe",
+		input:        []merkle.NodeFetch{f11, f12, f11, f11, f31, f11},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{n11, n12, n31},
+		result:       []storage.Node{n11, n12, n11, n11, n31, n11},
+	},
+	{
+		desc:         "storage fail",
+		input:        []merkle.NodeFetch{f11, f12, f11, f11, f31, f11},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{},
+		storageError: errors.New("storage"),
+	},
+	{
+		desc:         "storage wrong node",
+		input:        []merkle.NodeFetch{f11, f12, f11, f11, f31, f11},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{},
+		storageError: errors.New("storage"),
+	},
+}
+
+func TestRehasher(t *testing.T) {
+	for _, rehashTest := range rehashTests {
+		r := newRehasher()
```

Comment:
The rehasher is notionally separate API and I did consider pulling it out but it's not a massive amount of code. Most tests do call fetchNodesAndBuildProof. The dedup stuff has gone now.